### ISSUE TITLE
Type variable annotations for lambda functions with default arguments

### ISSUE THREAD

**Issue by dbrattli** [microsoft/pyright#2795]
**Describe the bug**

I cannot figure out how to handle type variable annotations for lambda functions with default arguments. Adding an issue since this might be a bug with pyright.

**To Reproduce**
Steps to reproduce the behavior.

Adding default values e.g `a=a` triggers the problem. 

```py
from typing import Callable

a: int = 10

lambda_151: Callable[[int, int], bool] = lambda y, a=a: a == y

print(lambda_151(20))
```

**Expected behavior**

My current thinking is that the code should type check. But currently reports multiple errors. Please advice.

<img width="898" alt="Screenshot 2022-01-03 at 09 50 49" src="https://user-images.githubusercontent.com/849479/147912936-07195087-d8c5-4f11-a022-c74eb03baee9.png">

(PS: in the image I tried to renaming `a` to `b`, but same errors)

**VS Code extension or command-line**

Pyright: v1.1.203 (vscode insiders)

**Comment by erictraut** [comment]
The `Callable` form doesn't provide a way to specify parameters with default argument values. (Nor does it allow for keyword parameters, *args, **kwargs, etc.). If you use `Callable` here, you will need to decide whether you want the resulting type to have one or two input parameters. It won't accept either one or two.

While `Callable` is quite limited in what it can represent, you can represent a much wider variety of function call signatures using a callback protocol. It would look something like this:

```python
class MyCallback(Protocol):
    def __call__(self, y: int, a: int = 10) -> bool:
        ...

lambda_153: MyCallback = lambda y, a=a: a == y
```

There is a bug here in that pyright is not performing bidirectional type inference for lambdas that have default parameter values. Since bidirectional type inference is the only means of inferring the parameter types in a lambda, this is problematic and is resulting in the "type is unknown" errors you're seeing. The fix for this is easy, and I'll make sure to include it in the next release.

**Comment by dbrattli** [comment]
Thanks for the good answer (and for fixing). I guess in most cases a typed lambda is probably not worth it compared to just defining a new typed function with some random name.

**Comment by erictraut** [comment]
This is included in pyright 1.1.204, which I just published. It will also be included in the next release of pylance.
